internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -78,11 +131,200 @@ func (r *NumaflowControllerRolloutReconciler) Reconcile(ctx context.Context, req
+
+	// TODO: add owner reference
+	if !controllerRollout.DeletionTimestamp.IsZero() {
+		numaLogger.Info("Deleting PipelineRollout")
+		if controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
+			controllerutil.RemoveFinalizer(controllerRollout, finalizerName)
+		}
+		return nil
+	}
+
+	// add Finalizer so we can ensure that we take appropriate action when CRD is deleted
+	if !controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
+		controllerutil.AddFinalizer(controllerRollout, finalizerName)
+	}
+
+	// apply controller
+	phase, err := r.sync(controllerRollout, namespace, numaLogger)
+	if err != nil {
+		return err
+	}
+
+	if phase != gitopsSyncCommon.OperationSucceeded {
+		return fmt.Errorf("sync operation is not successful")
+	}
+
+	//todo: controllerRollout.Status.MarkRunning() // should already be but just in case
+	return nil
+}
+
+// TODO: share sync logic with `syncer`
+func (r *NumaflowControllerRolloutReconciler) sync(
+	rollout *apiv1.NumaflowControllerRollout,
+	namespace string,
+	numaLogger *logger.NumaLogger,
+) (gitopsSyncCommon.OperationPhase, error) {
+
+	// Get the target manifests
+	version := rollout.Spec.Controller.Version
+	manifest := r.definitions[version]
+	numaLogger.Info("log manifest: " + manifest)
+	targetObjs, err := kubeUtil.SplitYAML([]byte(manifest))
+	if err != nil {
+		return gitopsSyncCommon.OperationError, err
+	}
+
+	var infoProvider kubeUtil.ResourceInfoProvider
+	clusterCache, err := r.stateCache.GetClusterCache()
+	infoProvider = clusterCache
+	if err != nil {
+		return gitopsSyncCommon.OperationError, err
+	}
+	liveObjByKey, err := r.stateCache.GetManagedLiveObjs(rollout.Name, targetObjs)
+	if err != nil {
+		return gitopsSyncCommon.OperationError, err
+	}
+	reconciliationResult := gitopsSync.Reconcile(targetObjs, liveObjByKey, namespace, infoProvider)
+
+	// Ignore `status` field for all comparison.
+	// TODO: make it configurable
+	overrides := map[string]sync.ResourceOverride{
+		"*/*": {
+			IgnoreDifferences: sync.OverrideIgnoreDiff{JSONPointers: []string{"/status"}}},
+	}
+
+	resourceOps, cleanup, err := r.getResourceOperations()
+	if err != nil {
+		return gitopsSyncCommon.OperationError, err
+	}
+	defer cleanup()
+
+	diffOpts := []diff.Option{
+		diff.WithLogr(*numaLogger.LogrLogger),
+		diff.WithServerSideDiff(true),
+		diff.WithServerSideDryRunner(diff.NewK8sServerSideDryRunner(resourceOps)),
+		diff.WithManager(common.SSAManager),
+		diff.WithGVKParser(clusterCache.GetGVKParser()),
+	}
+
+	diffResults, err := sync.StateDiffs(reconciliationResult.Target, reconciliationResult.Live, overrides, diffOpts)
+	if err != nil {
+		numaLogger.Error(err, "Error on comparing git sync state")
+		return gitopsSyncCommon.OperationError, err
+	}
+
+	opts := []gitopsSync.SyncOpt{
+		gitopsSync.WithLogr(*numaLogger.LogrLogger),
+		gitopsSync.WithOperationSettings(false, true, false, false),
+		gitopsSync.WithManifestValidation(true),
+		gitopsSync.WithPruneLast(false),
```

Comment:
any reason we shouldn't prune?

Reply:
we still `Prune`, just `PruneLast`, https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#prune-last

Reply:
oh, okay, thanks